File: Procosphere_Angular/src/app/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { environment } from '../../environments/environment';

export interface LoginRequest { email: string; password: string; }
export interface RegisterRequest {
  firstName: string; lastName: string; email: string; password: string;
  phone?: string; department?: string; company?: string;
  address?: string; avatarUrl?: string;
}
export type Role = 'ADMIN' | 'MANAGER' | 'USER' | 'SUPPLIER';
export interface AuthResponse {
  token: string; tokenType: string; userId: number;
  firstName: string; lastName: string; email: string; role: Role;
  mustChangePassword: boolean;
}
export interface ApiResponse<T> { success: boolean; message: string; data: T; }

export interface PasswordResetRequestResponse {
  id: number; userId: number; userName: string; userEmail: string;
  status: string; createdAt: string;
}

@Injectable({ providedIn: 'root' })
export class AuthService {
  private base = environment.apiUrl + '/api/auth';

  constructor(private http: HttpClient) {}

  login(req: LoginRequest): Observable<ApiResponse<AuthResponse>> {
    return this.http.post<ApiResponse<AuthResponse>>(`${this.base}/login`, req).pipe(
      tap(res => {
        if (res.success && res.data) {
          localStorage.setItem('token', res.data.token);
          localStorage.setItem('user', JSON.stringify(res.data));
          if (res.data.mustChangePassword) {
            localStorage.setItem('mustChangePassword', 'true');
          }
        }
      })
    );
  }

  register(req: RegisterRequest): Observable<ApiResponse<AuthResponse>> {
    return this.http.post<ApiResponse<AuthResponse>>(`${this.base}/register`, req);
  }

  logout(): void {
    localStorage.removeItem('token');
    localStorage.removeItem('user');
    localStorage.removeItem('mustChangePassword');
  }

  // --- Forgot Password ---

  forgotPassword(email: string): Observable<ApiResponse<null>> {
    return this.http.post<ApiResponse<null>>(`${this.base}/forgot-password`, { email });
  }

  loginByEmail(email: string): Observable<ApiResponse<AuthResponse>> {
    return this.http.post<ApiResponse<AuthResponse>>(`${this.base}/login-email`, { email }).pipe(
      tap(res => {
        if (res.success && res.data) {
          localStorage.setItem('token', res.data.token);
          localStorage.setItem('user', JSON.stringify(res.data));
          localStorage.setItem('mustChangePassword', 'true');
        }
      })
    );
  }

  setNewPassword(newPassword: string): Observable<ApiResponse<null>> {
    return this.http.post<ApiResponse<null>>(`${this.base}/set-password`, { newPassword }).pipe(
      tap(() => localStorage.removeItem('mustChangePassword'))
    );
  }

  // --- Admin: Password Reset Requests ---

  getResetRequests(): Observable<ApiResponse<PasswordResetRequestResponse[]>> {
    return this.http.get<ApiResponse<PasswordResetRequestResponse[]>>(
      `${this.base}/password-reset-requests`
    );
  }

  getPendingResetRequests(): Observable<ApiResponse<PasswordResetRequestResponse[]>> {
    return this.http.get<ApiResponse<PasswordResetRequestResponse[]>>(
      `${this.base}/password-reset-requests/pending`
    );
  }

  approveResetRequest(id: number): Observable<ApiResponse<null>> {
    return this.http.patch<ApiResponse<null>>(
      `${this.base}/password-reset-requests/${id}/approve`, {}
    );
  }

  rejectResetRequest(id: number): Observable<ApiResponse<null>> {
    return this.http.patch<ApiResponse<null>>(
      `${this.base}/password-reset-requests/${id}/reject`, {}
    );
  }

  // --- Helpers ---

  getToken(): string | null { return localStorage.getItem('token'); }

  isLoggedIn(): boolean { return !!this.getToken(); }

  mustChangePassword(): boolean {
    return localStorage.getItem('mustChangePassword') === 'true';
  }

  getCurrentUser(): AuthResponse | null {
    const u = localStorage.getItem('user');
    if (!u) return null;
    try { return JSON.parse(u); }
    catch { localStorage.removeItem('user'); return null; }
  }

  getCurrentUserId(): number | null {
    return this.getCurrentUser()?.userId ?? null;
  }
}

File: Backend/src/main/java/com/cts/mfrp/procuresphere/service/AuthService.java
package com.cts.mfrp.procuresphere.service;

import com.cts.mfrp.procuresphere.dto.request.LoginRequest;
import com.cts.mfrp.procuresphere.dto.request.RegisterRequest;
import com.cts.mfrp.procuresphere.dto.response.AuthResponse;
import com.cts.mfrp.procuresphere.dto.response.PasswordResetRequestResponse;
import com.cts.mfrp.procuresphere.exception.BadRequestException;
import com.cts.mfrp.procuresphere.exception.ResourceNotFoundException;
import com.cts.mfrp.procuresphere.model.PasswordResetRequest;
import com.cts.mfrp.procuresphere.model.Role;
import com.cts.mfrp.procuresphere.model.User;
import com.cts.mfrp.procuresphere.repository.PasswordResetRequestRepository;
import com.cts.mfrp.procuresphere.repository.UserRepository;
import com.cts.mfrp.procuresphere.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final JwtTokenProvider jwtTokenProvider;
    private final PasswordResetRequestRepository resetRequestRepository;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BadRequestException("Email '" + request.getEmail() + "' is already registered");
        }

        User user = User.builder()
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .phone(request.getPhone())
                .role(Role.USER)
                .department(request.getDepartment())
                .company(request.getCompany())
                .address(request.getAddress())
                .avatarUrl(request.getAvatarUrl())
                .rememberMe(false)
                .build();

        userRepository.save(user);
        String token = jwtTokenProvider.generateToken(user.getEmail());
        return buildAuthResponse(user, token, false);
    }

    public AuthResponse login(LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword()));

        String token = jwtTokenProvider.generateToken(authentication);
        User user = userRepository.findByEmail(request.getEmail()).orElseThrow();
        return buildAuthResponse(user, token, false);
    }

    // --- Forgot Password Flow ---

    @Transactional
    public void forgotPassword(String email) {
        if (email == null || !email.contains("@")) {
            throw new BadRequestException("Please provide a valid email address.");
        }
        User user = userRepository.findByEmail(email.trim())
                .orElse(null);
        // Return generic success even if email not found to avoid email enumeration
        if (user == null) return;

        // Check if there is already a pending request
        boolean alreadyPending = resetRequestRepository
                .findFirstByUserUserIdAndStatusOrderByCreatedAtDesc(user.getUserId(), "PENDING")
                .isPresent();
        if (alreadyPending) return; // silently ignore duplicate — don't reveal state

        PasswordResetRequest req = PasswordResetRequest.builder()
                .user(user)
                .status("PENDING")
                .build();
        resetRequestRepository.save(req);
    }

    @Transactional(readOnly = true)
    public List<PasswordResetRequestResponse> getAllResetRequests() {
        return resetRequestRepository.findAll().stream()
                .map(this::toResetResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<PasswordResetRequestResponse> getPendingResetRequests() {
        return resetRequestRepository.findByStatus("PENDING").stream()
                .map(this::toResetResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public void approvePasswordReset(Long requestId) {
        PasswordResetRequest req = resetRequestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Reset request not found: " + requestId));
        if (!"PENDING".equals(req.getStatus())) {
            throw new BadRequestException("Request is not in PENDING state.");
        }
        req.setStatus("APPROVED");
        resetRequestRepository.save(req);
    }

    @Transactional
    public void rejectPasswordReset(Long requestId) {
        PasswordResetRequest req = resetRequestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Reset request not found: " + requestId));
        if (!"PENDING".equals(req.getStatus())) {
            throw new BadRequestException("Request is not in PENDING state.");
        }
        req.setStatus("REJECTED");
        resetRequestRepository.save(req);
    }

    /**
     * Email-only login: allowed only if the user has an APPROVED password reset request.
     * Returns a JWT token with mustChangePassword = true.
     */
    @Transactional
    public AuthResponse loginByEmail(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("No account found with email: " + email));

        PasswordResetRequest req = resetRequestRepository
                .findFirstByUserUserIdAndStatusOrderByCreatedAtDesc(user.getUserId(), "APPROVED")
                .orElseThrow(() -> new BadRequestException(
                        "No approved password reset request found for this email. Please wait for admin approval."));

        // Mark it as USED so the user cannot log in this way again
        req.setStatus("USED");
        resetRequestRepository.save(req);

        String token = jwtTokenProvider.generateToken(user.getEmail());
        return buildAuthResponse(user, token, true);
    }

    /**
     * Set a new password after email-only login. Clears the mustChangePassword flag.
     */
    @Transactional
    public void setNewPassword(String userEmail, String newPassword) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + userEmail));
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }

    private AuthResponse buildAuthResponse(User user, String token, boolean mustChangePassword) {
        return AuthResponse.builder()
                .token(token)
                .tokenType("Bearer")
                .userId(user.getUserId())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .email(user.getEmail())
                .role(user.getRole())
                .mustChangePassword(mustChangePassword)
                .build();
    }

    private PasswordResetRequestResponse toResetResponse(PasswordResetRequest req) {
        User u = req.getUser();
        String name = (u.getFirstName() != null ? u.getFirstName() : "") + " " +
                      (u.getLastName()  != null ? u.getLastName()  : "");
        return PasswordResetRequestResponse.builder()
                .id(req.getId())
                .userId(u.getUserId())
                .userName(name.trim().isEmpty() ? u.getEmail() : name.trim())
                .userEmail(u.getEmail())
                .status(req.getStatus())
                .createdAt(req.getCreatedAt())
                .build();
    }
}
